test/IISIntegration.FunctionalTests/Utilities/TestServer.cs

```diff
@@ -0,0 +1,129 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Net.Http;
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Server.IntegrationTesting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+
+namespace IISIntegration.FunctionalTests.Utilities
+{
+    class TestServer: IDisposable, IStartup
+    {
+        private static readonly SemaphoreSlim WebCoreLock = new SemaphoreSlim(1, 1);
+
+        private static readonly int BasePort = 50690;
+        private static readonly Uri BaseUri = new Uri("http://localhost:" + BasePort);
+
+        private readonly TaskCompletionSource<object> _taskCompletionSource = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+        private readonly TaskCompletionSource<object> _shutdown = new TaskCompletionSource<object>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        private readonly RequestDelegate _app;
+        private readonly ILoggerFactory _loggerFactory;
+
+        public HttpClient HttpClient { get; }
+        public TestConnection CreateConnection() => new TestConnection(BasePort);
+
+        private IWebHost _host;
+
+        private TestServer(RequestDelegate app, ILoggerFactory loggerFactory)
+        {
+            _app = app;
+            _loggerFactory = loggerFactory;
+
+            HttpClient = new HttpClient(new LoggingHandler(new SocketsHttpHandler(), _loggerFactory.CreateLogger<TestServer>()))
+            {
+                BaseAddress = BaseUri
+            };
+        }
+
+        public static async Task<TestServer> Create(RequestDelegate app, ILoggerFactory loggerFactory)
+        {
+            await WebCoreLock.WaitAsync();
+            var server = new TestServer(app, loggerFactory);
+            server.Start();
+            await server.HttpClient.GetAsync("/start");
+            await server._taskCompletionSource.Task;
+            return server;
+        }
+
+        private void Start()
+        {
+            LoadLibrary(Environment.ExpandEnvironmentVariables(@"%windir%\system32\inetsrv\hwebcore.dll"));
```

Comment:
nit space.

Reply:
Where?

Reply:
Nvrm, ignoring whitespace changed the diff.